docs/design/2022-08-29-sync-stats-for-optimizer.md

```diff
@@ -0,0 +1,73 @@
+- Considering Sync Stats located in key path of executing a query, it should be efficient performance.
+- Sync Stats should expose its result if any error happened for easy diagnosis.
+
+## Proposal
+
+### Design
+
+To achieve the goals, `Sync Stats` should record the required columns and indices in each query and retrieved the stats before physical optimizing.
+Thus, in logical optimizing, we added 2 new rule `collectPredicateColumnsPoint` and `syncWaitStatsLoadPoint` to collect the required stats and retrieve the stats from storage.
+
+#### Collect the required columns and indices
+
+In `collectPredicateColumnsPoint` rule phase, we collect the required stats by transferring the plan tree, and extract the columns and indices from the conditions in each logical operator in the plan tree.
+
+For example, for the query `select a from t where b = 1 and c = 1` and execution its plan is like following:
+
+```sql 
++---------------------------+----------+-----------+---------------+----------------------------------+
+| id                        | estRows  | task      | access object | operator info                    |
++---------------------------+----------+-----------+---------------+----------------------------------+
+| Projection_4              | 0.01     | root      |               | test.t.a                         |
+| └─TableReader_7           | 0.01     | root      |               | data:Selection_6                 |
+|   └─Selection_6           | 0.01     | cop[tikv] |               | eq(test.t.b, 1), eq(test.t.c, 1) |
+|     └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo   |
++---------------------------+----------+-----------+---------------+----------------------------------+
+```
+
+The column `b` and `c` are maintained in the `Selection_6` and `collectPredicateColumnsPoint` will extract them in the operator so that we can know what's the required stats.
+After collecting the columns and indices from the queries, we will also check them whether they are necessary to retrieve the stats.
+For each column and index maintained in the tidb-server, tidb-server will also maintain the stats loading status in memory as `StatsLoadedStatus`.
+If the `StatsLoadedStatus` of column or index is full loaded, it will not to load stats during logical optimizing.
+
+#### Retrieve the stats
+
+After collecting the necessary columns and indices, tidb-server will collect the stats during `syncWaitStatsLoadPoint` phase.
+For each tidb-server, it will start several sub sync stats loader to load stats for columns and indices given by the above phase.
+As each query may send loading stats request to sync stats loader, the required columns and indices might be duplicated between each query.
+Thus in order to improve the performance for sync stats loader and reduce loading duplicated stats for columns and indices, each sync stats loader will check other loaders in order to avoid loading duplicated stats.
```

Comment:
```suggestion
Thus in order to improve the performance for sync stats loaders and reduce loading duplicated stats for columns and indices, each sync stats loader will check other loaders in order to avoid loading duplicated stats.
```